src/main/java/htsjdk/samtools/cram/structure/Slice.java

```diff
@@ -18,135 +18,585 @@
+        // Each Slice has 1 core data block, plus zero or more external data blocks.
+        // Since an embedded reference block is just stored as an external block, it is included in
+        // the external block count, and does not need to be counted separately.
+        return 1 + getSliceBlocks().getNumberOfExternalBlocks();
+    }
+
+    public void write(final int major, final OutputStream outputStream) {
+        setContentIDs(new int[getSliceBlocks().getNumberOfExternalBlocks()]);
+        final int i = 0;
+        for (final int id : getSliceBlocks().getExternalContentIDs()) {
+            getContentIDs()[i] = id;
+        }
+        // establish our header block, then write it out
+        sliceHeaderBlock = Block.createRawSliceHeaderBlock(createSliceHeaderBlockContent(major, this));
+        sliceHeaderBlock.write(major, outputStream);
+
+        // write the core and external blocks
+        getSliceBlocks().writeBlocks(major, outputStream);
     }
 
-    public ReferenceContext getReferenceContext() {
-        return referenceContext;
+    private static byte[] createSliceHeaderBlockContent(final int major, final Slice slice) {
+        final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        ITF8.writeUnsignedITF8(slice.getAlignmentContext().getReferenceContext().getReferenceContextID(), byteArrayOutputStream);
+        ITF8.writeUnsignedITF8(slice.getAlignmentContext().getAlignmentStart(), byteArrayOutputStream);
+        ITF8.writeUnsignedITF8(slice.getAlignmentContext().getAlignmentSpan(), byteArrayOutputStream);
+        ITF8.writeUnsignedITF8(slice.getNumberOfRecords(), byteArrayOutputStream);
+        LTF8.writeUnsignedLTF8(slice.getGlobalRecordCounter(), byteArrayOutputStream);
+        ITF8.writeUnsignedITF8(slice.getNumberOfBlocks(), byteArrayOutputStream);
+
+        slice.setContentIDs(new int[slice.getSliceBlocks().getNumberOfExternalBlocks()]);
+        int i = 0;
+        for (final int id : slice.getSliceBlocks().getExternalContentIDs()) {
+            slice.getContentIDs()[i++] = id;
+        }
+        CramIntArray.write(slice.getContentIDs(), byteArrayOutputStream);
+        ITF8.writeUnsignedITF8(slice.getEmbeddedReferenceContentID(), byteArrayOutputStream);
+        try {
+            byteArrayOutputStream.write(slice.getReferenceMD5() == null ? new byte[16] : slice.getReferenceMD5());
+        } catch (final IOException e) {
+            throw new RuntimeIOException(e);
+        }
+
+        if (major >= CramVersions.CRAM_v3.major) {
+            if (slice.getSliceTags() != null) {
+                final BinaryCodec binaryCoded = new BinaryCodec(byteArrayOutputStream);
+                final BinaryTagCodec binaryTagCodec = new BinaryTagCodec(binaryCoded);
+                SAMBinaryTagAndValue samBinaryTagAndValue = slice.getSliceTags();
+                do {
+                    log.debug("Writing slice tag: " + SAMTag.makeStringTag(samBinaryTagAndValue.tag));
```

Comment:
this log message does calculations every time, wrap it in an if statement

Reply:
Removed completely.